basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java

```diff
@@ -442,6 +451,98 @@ protected void handleProxy(HttpServletRequest request, HttpServletResponse respo
+
+		String callback = request.getParameter("callback");
+		if ( callback == null ) {
+			LTI13Util.return400(response, "Missing callback parameter");
+			return;
+		}
+
+		Session sess = SessionManager.getCurrentSession();
+		if ( sess == null || sess.getUserId() == null ) {
+			response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+
+		Map<String, Object> content = loadContentCheckSignature(signed_placement, response);
+		if (content == null) {
+			response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+
+		Site site = loadSiteFromContent(content, signed_placement, response);
+		if (site == null) {
+			response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+
+		Long toolKey = getLongKey(content.get(LTIService.LTI_TOOL_ID));
+		if (toolKey < 0 ) {
+			log.error("Content / Tool invalid content={} tool={}", content.get(LTIService.LTI_ID), toolKey);
+			LTI13Util.return400(response, "Content / Tool mismatch");
+			return;
+		}
+
+		Map<String, Object> tool = ltiService.getToolDao(toolKey, site.getId());
+		if (tool == null) {
+			log.error("Could not load tool={}", toolKey);
+			LTI13Util.return400(response, "Missing tool");
+			return;
+		}
+
+		String platform_public = (String) tool.get(LTIService.LTI13_PLATFORM_PUBLIC);
+		String platform_private = SakaiBLTIUtil.decryptSecret((String) tool.get(LTIService.LTI13_PLATFORM_PRIVATE));
+
+		Key privateKey = LTI13Util.string2PrivateKey(platform_private);
+		Key publicKey = LTI13Util.string2PublicKey(platform_public);
+
+		String kid = LTI13KeySetUtil.getPublicKID(publicKey);
+
+		String context_id = site.getId();
+
+		String user_id = sess.getUserId();
+		String subject = SakaiBLTIUtil.getSubject(user_id, context_id);
+
+		try {
+			Long issued = new Long(System.currentTimeMillis() / 1000L);
+			String body =
+				"{\n" +
+				"\"iss\": \""+SakaiBLTIUtil.getOurServerUrl()+"\",\n" +
+				"\"aud\": \""+SakaiBLTIUtil.getOurServerUrl()+"\",\n" +
+				"\"exp\": \""+(issued+3600L)+"\",\n" +
+				"\"user_id\": \""+user_id+"\",\n" +
+				"\"context_id\": \""+context_id+"\",\n" +
+				"\"sub\": \""+subject+"\"\n" +
+				"}";
+
+			// http://javadox.com/io.jsonwebtoken/jjwt/0.4/io/jsonwebtoken/JwtBuilder.html
+			String jws = Jwts.builder()
+				.setHeaderParam("kid", kid)
+				.setPayload(body)
+				.signWith(privateKey)
+				.compact();
+
+			// https://stackoverflow.com/questions/3324717/sending-http-post-request-in-java
+			byte [] bytes = jws.getBytes();
+			java.net.URL url = new java.net.URL(callback);
+			java.net.HttpURLConnection con = (java.net.HttpURLConnection) url.openConnection();
+			con.setRequestMethod("POST");
+	        con.setDoOutput(true);
+			con.setFixedLengthStreamingMode(bytes.length);
+			con.setRequestProperty( "Content-Type", APPLICATION_JWT );
+	        con.connect();
+			OutputStream os = con.getOutputStream();
+			os.write(bytes);
+			os.flush();
+			os.close();
```

Comment:
I would move the output stream to a finally block or use the try with resources.